rules_gapic/java/java_gapic_pkg.bzl

```diff
@@ -354,6 +256,84 @@ def _java_gapic_gradle_pkg(
+    resource_target_name = "%s-resources" % name
+    java_gapic_build_configs_pkg(
+        name = resource_target_name,
+        deps = deps,
+        templates = {
+            Label("//rules_gapic/java:resources/gradle/assembly.gradle.tmpl"): "build.gradle",
+            Label("//rules_gapic/java:resources/gradle/settings.gradle.tmpl"): "settings.gradle",
+        },
+    )
+
+    pkg_tar(
+        name = name,
+        extension = "tar.gz",
+        deps = [
+            Label("//rules_gapic/java:gradlew"),
+            resource_target_name,
+        ] + deps,
+        package_dir = name,
         visibility = visibility,
     )
+
+def _construct_extra_deps(scope_to_deps, versions_map):
+    label_name_to_maven_artifact = {
+        "iam_policy_proto": "maven.com_google_api_grpc_proto_google_iam_v1",
+        "iam_java_proto": "maven.com_google_api_grpc_proto_google_iam_v1",
+        "iam_java_grpc": "maven.com_google_api_grpc_grpc_google_iam_v1",
+        "iam_policy_java_grpc": "maven.com_google_api_grpc_grpc_google_iam_v1",
+    }
+    extra_deps = {}
+    for scope, deps in scope_to_deps.items():
+        for dep in deps:
+            if _is_java_dependency(dep):
+                for f in dep[JavaInfo].transitive_deps.to_list():
+                    maven_artifact = label_name_to_maven_artifact.get(f.owner.name)
+                    if maven_artifact:
+                        key = "{{%s}}" % maven_artifact
+                        if not extra_deps.get(key):
+                            extra_deps[key] = "%s '%s'" % (scope, versions_map[key])
+            elif _is_gapic_pkg_dependency(dep):
+                dep_file = dep.files.to_list()[0]
+                dep_name = dep_file.basename
+                for ext in (".tar.gz", ".gz", ".tgz"):
+                    if dep_name.endswith(ext):
+                        dep_name = dep_name[:-len(ext)]
+                        break
+                key = "{{%s}}" % dep_name
+                if not extra_deps.get(key):
+                    extra_deps[key] = "%s project(':%s')" % (scope, dep_name)
+
+    return "\n  ".join(extra_deps.values())
+
+def _ends_with(s, sub_s):
+    len_s = len(s)
+    len_sub_s = len(sub_s)
+    return len_s >= len_sub_s and s.rfind(sub_s, len_s - len_sub_s) >= 0
+
+def _put_dep_in_a_bucket(dep, dep_bucket, processed_deps):
+    if processed_deps.get(dep):
+        return
+    dep_bucket.append(dep)
+    processed_deps[dep] = True
+
+def _is_java_dependency(dep):
+    return JavaInfo in dep
+
+def _is_source_dependency(dep):
+    return _is_java_dependency(dep) and hasattr(dep[JavaInfo], "source_jars") and dep.label.package != "jar"
+
+def _is_proto_dependency(dep):
+    return ProtoInfo in dep
+
+def _is_gapic_pkg_dependency(dep):
+    files_list = dep.files.to_list()
+    if not files_list or len(files_list) != 1:
+        return False
+    return files_list[0].extension in (".tar.gz", "gz", "tgz")
```

Comment:
This list of file compression extensions appears twice in this file. Should we pull it out into a variable like `_SUPPORTED_EXTENSIONS`?

This one does't have a `.` prefix on `gz` and `tgz` like the other list does, is that a real difference?

Reply:
Oh, I think `.` is my mistake, `extension` property does not have the dot, so it should not be in the string literal. 

You are right that there is a lot of duplication here. I cleaned up this whole logic.